### DISCUSSION TITLE
Render TikZ diagrams in Quarto

### DISCUSSION THREAD

**Discussion by danmackinlay** [quarto-dev/quarto-cli#8925] [Show and Tell]
### Description

This is part show-and-tell and part Q&A I guess ;-)

Announcing [TikZ for quarto](https://github.com/danmackinlay/quarto_tikz).

<img width="870" alt="Screenshot 2024-02-29 at 08 38 06" src="https://github.com/quarto-dev/quarto-cli/assets/21740/c658d218-8c09-4f40-ae09-53d41f891495">

TikZ is an time-honored (i.e. old) graphics language without the hype of [D2](https://d2lang.com/) or [mermaid](https://mermaid.js.org/)
However, after drawing a *lot* of diagrams where I needed to have specific control of the layout but where I didn't want to spend my entire time calculating the layout manually, I've decided it is the best diagram tool in the category of  **[diagram tools that automate the right amount of stuff](https://danmackinlay.name/notebook/diagrams_procedural.html#tikz).**

The only problem is: it is part of the TeX ecosystem and doesn't support SVG or HTML. Hence the new filter.
[kitr already solved that](https://github.com/yihui/knitr/blob/master/R/engine.R#L348) though, and  [data-intuitive/quarto-d2/](https://github.com/data-intuitive/quarto-d2/blob/main/_extensions/d2/d2.lua) showed us how to use filters to put new graphics formats in quarto.
After spending 2 days of my life getting this working, I found that [there is even a worked example of a tikz filter in pandoc itself](https://pandoc.org/lua-filters.html#building-images-with-tikz).

Anyway, this thing works now and I'm using it.

There are still outstanding problems, though; I'll open a Q&A thread about those.

**Comment by danmackinlay** [comment]
The outstanding problem is essentially that while this filter renders an image successfully, it doesn't manage to turn it into a figure, or give it an id, so it cannot be cross-referenced. [see this discussion](https://github.com/quarto-dev/quarto-cli/discussions/8926)

**Reply by cscheid** [reply]
> The outstanding problem is essentially that while this filter renders an image successfully, it doesn't manage to turn it into a figure, or give it an id, so it cannot be cross-referenced. https://github.com/quarto-dev/quarto-cli/discussions/8926

That kind of thing is why we came up with the div syntax for cross-references and the new, more permissive structure of Quarto 1.4 crossref's system. What happens if you try the following?

````

::: {#fig-tikz}

```{.tikz ...}
...
```

This is a caption
:::

See @fig-tikz.
````

**Comment by mcanouil** [comment]
Were you aware of <https://github.com/pandoc-ext/diagram>?

**Reply by danmackinlay** [reply]
No! 🤦🏻

**Reply by danmackinlay** [reply]
Here is a pull request to list it with the other extensions: https://github.com/quarto-dev/quarto-web/pull/1035

**Reply by danmackinlay** [reply]
Hmm, https://github.com/pandoc-ext/diagram is generally more powerful and capable that my paltry code except it seems to run into the problem that it emits Figure elements, which if I understand correctly will [make it hard to crossref](https://github.com/quarto-dev/quarto-cli/discussions/8926#discussioncomment-8624722). I'll do some experiments.

**Reply by cscheid** [reply]
> Hmm, https://github.com/pandoc-ext/diagram is generally more powerful and capable that my paltry code except it seems to run into the problem that it emits Figure elements, which if I understand correctly will https://github.com/quarto-dev/quarto-cli/discussions/8926#discussioncomment-8624722. I'll do some experiments.

You have it right. The same div syntax should work in this case, but our output might be a bit funky with nested `<figure>` elements.

**Comment by stevengogogo** [comment]
This works for me: https://stackoverflow.com/questions/71760913/using-tikz-in-quarto-presentation

**Comment by allefeld** [comment]
Not exactly the same, but my [pyTikZ](https://github.com/allefeld/pytikz) project provides a Python interface to create TikZ graphics. It is a bit older, from before I started using Quarto, so it is mainly tested in Jupyter notebooks, but in should work in Quarto, too.
